kedro/io/kedro_data_catalog.py

```diff
@@ -0,0 +1,335 @@
+"""``KedroDataCatalog`` stores instances of ``AbstractDataset`` implementations to
+provide ``load`` and ``save`` capabilities from anywhere in the program. To
+use a ``KedroDataCatalog``, you need to instantiate it with a dictionary of datasets.
+Then it will act as a single point of reference for your calls, relaying load and
+save functions to the underlying datasets.
+"""
+
+from __future__ import annotations
+
+import copy
+import difflib
+import logging
+import re
+from typing import Any
+
+from kedro.io.catalog_config_resolver import CatalogConfigResolver, Patterns
+from kedro.io.core import (
+    AbstractDataset,
+    AbstractVersionedDataset,
+    DatasetAlreadyExistsError,
+    DatasetError,
+    DatasetNotFoundError,
+    Version,
+    generate_timestamp,
+)
+from kedro.io.memory_dataset import MemoryDataset
+from kedro.utils import _format_rich, _has_rich_handler
+
+
+class KedroDataCatalog:
```

Comment:
Should this be: `class KedroDataCatalog(CatalogProtocol)` ?

Reply:
We only need it in case we share some logic between the implementations which we intentionally don't want to do to keep all the implementations independent from the protocol class.

> while it’s possible to subclass a protocol explicitly, it’s not necessary to do so for the sake of type-checking

https://peps.python.org/pep-0544/#explicitly-declaring-implementation

Reply:
I don't mind, but the point was to clearly show that we do not have a shared logic and addition of new catalog does not require an explicit declaration.